onion_router/onion_router.py

```diff
@@ -19,21 +21,103 @@
 
 
 
-#       RUN DUMMY SERVER
+#       GLOBALS
 ########################################################
 
 PORT = args.port
 HOST = ""
-SOCKET_LISTEN = socket(AF_INET, SOCK_STREAM)
-SOCKET_LISTEN.bind((HOST,PORT))
-SOCKET_LISTEN.listen(1)
-
-try:
-    while 1:
-        conn_socket, client_addr = SOCKET_LISTEN.accept()
-        msg = conn_socket.recv(2048)
-        mod_msg = msg.upper()
-        conn_socket.send(mod_msg)
-        conn_socket.close()
-except KeyboardInterrupt:
-    SOCKET_LISTEN.close()
+LISTENER_SOCKET = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+LISTENER_SOCKET.bind((HOST,PORT))
+LISTENER_SOCKET.listen(50)
+
+
+
+#       MAIN
+########################################################
+
+def main():
+    try:
+        while 1:
+            # accept connection from previous ('backward') node in path
+            back_conn, client_addr = LISTENER_SOCKET.accept()
+            print('Connected to: {}:{}'.format(client_addr[0], str(client_addr[1])))
+
+            # new thread for self-setup as onion_router
+            t = threading.Thread(target=two_way_setup, args=(back_conn,))
+            t.start()
+
+    except (socket.error, KeyboardInterrupt):
+        LISTENER_SOCKET.close()
+
+
+
+#       SETUP TWO TCP CONNECTIONS
+########################################################
+
+# need to establish symkey with back_conn, connect to forw_conn
+def two_way_setup(back_conn):
+    msg = back_conn.recv(2048).decode('utf-8').rstrip()
+
+    # (TODO:decrypt data, initialize+respond with symkey. for now send 'ACK')
+    back_conn.send("ACK".encode('utf-8'))
+
+    # Wait for first-ever onion from back_conn
+    print('[Onion] Waiting for data onion...')
+    msg = back_conn.recv(2048).decode('utf-8').rstrip()
+    print('[Onion] Got data onion.')
+
+    # (TODO:decrypt)
+
+    # parse onion to find out who to send to and what to send
+    msg_dict = json.loads(msg)
```

Comment:
@stumash @DestructiveReasoning 
You are probably already aware, but for the sake of keeping up to date:

With respect to Harley's asciinema video, we don't need to escape internal quotes.

Because of this line, I am able to communicate from src to dst with 3 intermediate nodes **without** the need to escaping the internal quotations:

> Cebolla > nc cs-1 5555
setup_vs_cs-1
ACK{"addr":"cs-2","data":"setup_vs_cs-2"}
ACK{"addr":"cs-2","data":{"addr":"cs-3","data":"setup_vs_cs-3"}}
ACK{"addr":"cs-2","data":{"addr":"cs-3","data":{"addr":"cs-4","data":"nepe"}}}
NEPE

Reply:
so happy you saw my `json.dumps()` calls.  I put them in after he recorded.  We previously were just doing `conn_send(str(msg_data).encode('utf-8'))` and then we had to escape the quotes for some reason.